File: src/etch/bindings/csharp/support/test/TestRemoteBase.cs
///$Id$
///
///Created by Champakesan, Badri Narayanan on Jul 02, 2007.
///
///Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.

using System;
using NUnit.Framework;
using Etch.Msg;
using Etch.Support;
using System.Collections.Generic;
//using Etch.Support.Test;

namespace Etch.Support.Test
{

    [TestFixture]
    public class TestRemoteBase
    {
        public void Init()
        {
            svc.xmsg = null;
            svc.xresponseType = null;
            svc.xresponseField = null;
            svc.xtimeout = 0;
        }

        private static MyDeliveryService svc = new MyDeliveryService(); //declared static
        //private static DefaultValueFactory vf = new DefaultValueFactory();//declared static
        private static MyDefaultValueFactory vf = new MyDefaultValueFactory();
        private RemoteBase<DefaultValueFactory> rb = new RemoteBase<DefaultValueFactory>(svc, vf);
    	private XType mt = new XType( "foo" );
    	private XType rmt = new XType( "bar" );
    	private Field rmf = new Field( "baz" );
	
	    [Test]
	    public void NewMessage()
	    {
		    Message msg = rb._NewMessage( mt );
		    msg.CheckType( mt );
		    Assert.AreSame( vf, msg._vf );
		    Assert.IsNull( svc.what );
		    Assert.IsNull( svc.xmsg );
		    Assert.IsNull( svc.xresponseType );
		    Assert.IsNull( svc.xresponseField );
		    Assert.AreEqual( 0, svc.xtimeout );
	    }
    	
	    [Test]
	    public void Send() 
	    {
		    Message msg = rb._NewMessage( mt );
    		
		    rb._svc.Send( msg );
		    Assert.IsTrue( svc.what.Equals(What.SEND));
		    Assert.AreSame( svc.xmsg, msg );
		    Assert.IsNull( svc.xresponseType );
		    Assert.IsNull( svc.xresponseField );
		    Assert.AreEqual( 0, svc.xtimeout );
	    }
    	
        [Test]
	    public void Call() 
	    {
		    Message msg = rb._NewMessage( mt );

            Object result = rb._svc.EndCall( rb._svc.BeginCall( msg ), rmt, rmf, 98 );

		    Assert.IsTrue( svc.what.Equals(What.CALL));
		    Assert.AreSame( svc.xmsg, msg );
		    Assert.AreSame( svc.xresponseType, rmt );
		    Assert.AreSame( svc.xresponseField, rmf );
		    Assert.AreEqual( svc.xtimeout, 98 );
		    Assert.AreEqual( result, 23 );
	    }
    	
	    public enum What { SEND, CALL, SHUTDOWN, BEGINCALL };
    	
	    public class MyDeliveryService : DeliveryService
	    {

            public Enum what;
            
            public void Init()
            {
                what = null;
            }

		    public Message xmsg;
		    public XType xresponseType;
		    public Field xresponseField;
		    public int xtimeout;
            public Mailbox xmb;
    		
		    public void Send( Message msg ) 
		    {
			    Assert.IsNull( what );
			    what = What.SEND;
			    this.xmsg = msg;
		    }

            public Mailbox BeginCall( Message msg )
            {
                Init();
                Assert.IsNull( what );
                what = What.BEGINCALL;
                this.xmsg = msg;
                this.xmb = new Etch.Transport.PlainMailbox( null, 0L, 0, 0, 1 );
                return this.xmb;
            }

            public Object EndCall( Mailbox mb, XType responseType,
			                    Field responseField, int timeout ) 
		    {
                Assert.AreEqual( What.BEGINCALL, what );
                Assert.AreEqual( xmb, mb );
			    what = What.CALL;
                xmb = null;
			    this.xresponseType = responseType;
			    this.xresponseField = responseField;
			    this.xtimeout = timeout;
			    return 23;
		    }

            public void TransportNotify( Object eventObj )
            {
                // ignore
            }

            public Object TransportQuery( Object query )
            {
                // ignore
                return null;
            }

            public void TransportControl( Object control, Object value )
            {
                // ignore
            }
	    }

        /// <summary>
        /// A fake implementation of DefaultValueFactory for testing
        /// </summary>
        /// 
	    public class MyDefaultValueFactory : DefaultValueFactory
	    {
		    private readonly static TypeMap types = new TypeMap();

		    private readonly static FieldMap fields = new FieldMap();

		    static MyDefaultValueFactory()
		    {
			    DefaultValueFactory.Init( types, fields );
		    }

		    /**
		     * Constructs the ValueFactoryFake.
		     */
		    public MyDefaultValueFactory()
		    {
			    // nothing to do.
		    }

		    public override Field GetField( int? id )
		    {
			    return fields.Get( id );

		    }

            public override Field GetField( String name )
		    {
			    return fields.Get( name );
		    }

            public override ICollection<Field> GetFields()
		    {
			    return fields.Values();
		    }

            public override XType GetType( String name )
		    {
			    return types.Get( name );
		    }

            public override ICollection<XType> GetTypes()
		    {
			    return types.Values();
		    }

            public override XType GetType( int? id )
		    {
			    return types.Get( id );
		    }
	    }
  }
}

File: src/etch/bindings/java/support/test/TestRemoteBase.java
package etch.bindings.java.support.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;

import org.junit.Test;

import etch.bindings.java.msg.Field;
import etch.bindings.java.msg.Message;
import etch.bindings.java.msg.Type;
import etch.bindings.java.support.DefaultValueFactory;
import etch.bindings.java.support.DeliveryService;
import etch.bindings.java.support.Mailbox;
import etch.bindings.java.support.RemoteBase;
import etch.bindings.java.transport.PlainMailbox;

/**
 * Test of RemoteBase.
 */
public class TestRemoteBase
{
	private MyDeliveryService svc = new MyDeliveryService();
	private DefaultValueFactory vf = new DefaultValueFactory();
	private RemoteBase<DefaultValueFactory> rb = new RemoteBase<DefaultValueFactory>( svc, vf );
	private Type mt = new Type( "foo" );
	private Type rmt = new Type( "bar" );
	private Field rmf = new Field( "baz" );
	
	/** @throws Exception */
	@Test
	public void newMessage() throws Exception
	{
		Message msg = rb._newMessage( mt );
		msg.checkType( mt );
		assertSame( vf, msg.vf() );
		assertNull( svc.what );
		assertNull( svc.xmsg );
		assertNull( svc.xresponseType );
		assertNull( svc.xresponseField );
		assertNull( svc.xtimeout );
	}

	/** @throws Exception */
	@Test
	public void send() throws Exception
	{
		Message msg = rb._newMessage( mt );
		
		rb._svc.send( msg );
		
		assertSame( svc.what, What.SEND );
		assertSame( svc.xmsg, msg );
		assertNull( svc.xresponseType );
		assertNull( svc.xresponseField );
		assertNull( svc.xtimeout );
	}

	/** @throws Exception */
	@Test
	public void voidcall() throws Exception
	{
		Message msg = rb._newMessage( mt );
		
		rb._svc.endvoidcall( rb._svc.begincall( msg ), rmt, rmf, 99 );
		
		assertSame( svc.what, What.VOIDCALL );
		assertSame( svc.xmsg, msg );
		assertSame( svc.xresponseType, rmt );
		assertSame( svc.xresponseField, rmf );
		assertEquals( svc.xtimeout, 99 );
	}

	/** @throws Exception */
	@Test
	public void call() throws Exception
	{
		Message msg = rb._newMessage( mt );
		
		Object result = rb._svc.endcall( rb._svc.begincall( msg ), rmt, rmf, 98 );
		
		assertSame( svc.what, What.CALL );
		assertSame( svc.xmsg, msg );
		assertSame( svc.xresponseType, rmt );
		assertSame( svc.xresponseField, rmf );
		assertEquals( svc.xtimeout, 98 );
		assertEquals( result, 23 );
	}
	
	/**
	 * testing events.
	 */
	enum What
	{
		/** send was called */
		SEND,
		/** voidcall was called */
		VOIDCALL,
		/** call was called */
		CALL,
		/** begincall was called */
		BEGINCALL
	};
	
	/**
	 * DeliveryService for testing.
	 */
	public static class MyDeliveryService implements DeliveryService
	{
		/** event */
		public What what;
		/** message sent */
		public Message xmsg;
		/** type of response received */
		public Type xresponseType;
		/** field of response received */
		public Field xresponseField;
		/** timeout received */
		public Integer xtimeout;
		/** mailbox to return */
		public Mailbox xmb;
		
		public void send( Message msg ) throws Exception
		{
			assertNull( what );
			what = What.SEND;
			this.xmsg = msg;
		}
		
		public Mailbox begincall( Message msg )
		{
			assertNull( what );
			what = What.BEGINCALL;
			this.xmsg = msg;
			this.xmb = new PlainMailbox( null, 0L, 0, 0, 1 );
			return this.xmb;
		}

		public void endvoidcall( Mailbox mb, Type responseType,
			Field responseField, int timeout ) throws Exception
		{
			assertSame( What.BEGINCALL, what );
			assertSame( xmb, mb );
			what = What.VOIDCALL;
			xmb = null;
			this.xresponseType = responseType;
			this.xresponseField = responseField;
			this.xtimeout = timeout;
		}
		
		public Object endcall( Mailbox mb, Type responseType,
			Field responseField, int timeout ) throws Exception
		{
			assertSame( What.BEGINCALL, what );
			assertSame( xmb, mb );
			what = What.CALL;
			xmb = null;
			this.xresponseType = responseType;
			this.xresponseField = responseField;
			this.xtimeout = timeout;
			return 23;
		}

		public void transportControl( Object control, Object value ) throws Exception
		{
			// TODO Auto-generated method stub
			
		}

		public void transportNotify( Object event ) throws Exception
		{
			// TODO Auto-generated method stub
			
		}

		public Object transportQuery( Object query ) throws Exception
		{
			// TODO Auto-generated method stub
			return null;
		}
	}
}
